stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+}
+
+impl NakamotoTenureDownloaderSet {
+    pub fn new() -> Self {
+        Self {
+            downloaders: vec![],
+            peers: HashMap::new(),
+            completed_tenures: HashSet::new(),
+        }
+    }
+
+    /// Assign the given peer to the given downloader state machine.  Allocate a slot for it if
+    /// needed.
+    fn add_downloader(&mut self, naddr: NeighborAddress, downloader: NakamotoTenureDownloader) {
+        test_debug!(
+            "Add downloader for tenure {} driven by {}",
+            &downloader.tenure_id_consensus_hash,
+            &naddr
+        );
+        if let Some(idx) = self.peers.get(&naddr) {
+            self.downloaders[*idx] = Some(downloader);
+        } else {
+            self.downloaders.push(Some(downloader));
+            self.peers.insert(naddr, self.downloaders.len() - 1);
+        }
+    }
+
+    /// Does the given neighbor have an assigned downloader state machine?
+    pub(crate) fn has_downloader(&self, naddr: &NeighborAddress) -> bool {
+        let Some(idx) = self.peers.get(naddr) else {
+            return false;
+        };
+        let Some(downloader_opt) = self.downloaders.get(*idx) else {
+            return false;
+        };
+        downloader_opt.is_some()
+    }
+
+    /// Drop the downloader associated with the given neighbor, if any.
+    pub fn clear_downloader(&mut self, naddr: &NeighborAddress) {
+        let Some(index) = self.peers.remove(naddr) else {
+            return;
+        };
+        self.downloaders[index] = None;
+    }
+
+    /// How many downloaders are there?
+    pub fn num_downloaders(&self) -> usize {
+        self.downloaders
+            .iter()
+            .fold(0, |acc, dl| if dl.is_some() { acc + 1 } else { acc })
+    }
+
+    /// How many downloaders are there, which are scheduled?
+    pub fn num_scheduled_downloaders(&self) -> usize {
+        let mut cnt = 0;
+        for (_, idx) in self.peers.iter() {
+            if let Some(Some(_)) = self.downloaders.get(*idx) {
+                cnt += 1;
+            }
+        }
+        cnt
+    }
+
+    /// Add a sequence of (address, downloader) pairs to this downloader set.
+    pub(crate) fn add_downloaders(
+        &mut self,
+        iter: impl IntoIterator<Item = (NeighborAddress, NakamotoTenureDownloader)>,
+    ) {
+        for (naddr, downloader) in iter {
+            if self.has_downloader(&naddr) {
+                test_debug!("Already have downloader for {}", &naddr);
+                continue;
+            }
+            self.add_downloader(naddr, downloader);
+        }
+    }
+
+    /// Count up the number of in-flight messages, based on the states of each instantiated
+    /// downloader.
+    pub fn inflight(&self) -> usize {
+        let mut cnt = 0;
+        for downloader_opt in self.downloaders.iter() {
+            let Some(downloader) = downloader_opt else {
+                continue;
+            };
+            if downloader.idle {
```

Comment:
Would we ever want to make a helper function for a downloader to check if its inflight? such as pub is_inflight() like how you have "is_waiting" which wraps the rest of the calls?

Reply:
I suppose, but that function would only return `self.idle` (which takes more keystrokes than just loading that public instance variable).